g++ 4.8 still confused about shared_ptr<T> being constructible from SEXP

This looks like a bug of the compiler IIUC. https://github.com/apache/arrow/pull/7819/checks?check_run_id=997448752

```
array_from_vector.cpp: In function ‘std::shared_ptr<arrow::DataType> arrow::r::InferArrowTypeFromVector(SEXP) [with int VectorType = 4; SEXP = SEXPREC*]’:
3602
array_from_vector.cpp:1198:58: error: call of overloaded ‘as_cpp(SEXPREC*&)’ is ambiguous
3603
     return cpp11::as_cpp<std::shared_ptr<arrow::Array>>(x)->type();
3604
                                                          ^
3605
array_from_vector.cpp:1198:58: note: candidates are:
3606
In file included from /opt/R/4.0.2/lib/R/library/cpp11/include/cpp11.hpp:5:0,
3607
                 from ././arrow_cpp11.h:36,
3608
                 from ./arrow_types.h:22,
3609
                 from array_from_vector.cpp:20:
3610
/opt/R/4.0.2/lib/R/library/cpp11/include/cpp11/as.hpp:66:41: note: cpp11::enable_if_constructible_from_sexp<T, T> cpp11::as_cpp(SEXP) [with T = std::shared_ptr<arrow::Array>; cpp11::enable_if_constructible_from_sexp<T, T> = std::shared_ptr<arrow::Array>; SEXP = SEXPREC*]
3611
 enable_if_constructible_from_sexp<T, T> as_cpp(SEXP from) {
3612
                                         ^
3613
In file included from ./arrow_types.h:22:0,
3614
                 from array_from_vector.cpp:20:
3615
././arrow_cpp11.h:215:25: note: cpp11::enable_if_shared_ptr<T> cpp11::as_cpp(SEXP) [with T = std::shared_ptr<arrow::Array>; cpp11::enable_if_shared_ptr<T> = std::shared_ptr<arrow::Array>; SEXP = SEXPREC*]
3616
 enable_if_shared_ptr<T> as_cpp(SEXP from) {
3617
                         ^
3618
```

--- a/inst/include/cpp11/as.hpp
+++ b/inst/include/cpp11/as.hpp
@@ -31,7 +31,11 @@ struct is_smart_ptr<std::weak_ptr<T>> : std::true_type {};
 
 template <typename T, typename R = void>
 using enable_if_constructible_from_sexp =
-    enable_if_t<std::is_class<T>::value && std::is_constructible<T, SEXP>::value, R>;
+    enable_if_t<
+      !is_smart_ptr<T>::value && // workaround for gcc 4.8
+      std::is_class<T>::value && std::is_constructible<T, SEXP>::value,
+      R
+    >;
 
 template <typename T, typename R = void>
 using enable_if_is_sexp = enable_if_t<std::is_same<T, SEXP>::value, R>;
@@ -75,7 +79,7 @@ inline bool is_convertable_without_loss_to_integer(double value) {
   return std::modf(value, &int_part) == 0.0;
 }
 
-template <typename T, typename = enable_if_t<!is_smart_ptr<T>::value>>
+template <typename T>
 enable_if_constructible_from_sexp<T, T> as_cpp(SEXP from) {
   return T(from);
 }

--- a/inst/include/cpp11/as.hpp
+++ b/inst/include/cpp11/as.hpp
@@ -17,7 +17,7 @@ template <typename T>
 using decay_t = typename std::decay<T>::type;
 
 template <typename T>
-struct is_smart_ptr : std::false_type {};;
+struct is_smart_ptr : std::false_type {};
 
 template <typename T>
 struct is_smart_ptr<std::shared_ptr<T>> : std::true_type{};

--- a/inst/include/cpp11/as.hpp
+++ b/inst/include/cpp11/as.hpp
@@ -1,9 +1,10 @@
 #pragma once
 
-#include <cmath>             // for modf
-#include <initializer_list>  // for initializer_list
-#include <string>            // for string, basic_string
-#include <type_traits>       // for decay, enable_if, is_same, is_convertible
+#include <cmath>              // for modf
+#include <initializer_list>   // for initializer_list
+#include <string>             // for string, basic_string
+#include <type_traits>        // for decay, enable_if, is_same, is_convertible
+#include <memory>             // for std::shared_ptr, std::weak_ptr, std::unique_ptr
 
 #include "cpp11/R.hpp"        // for SEXP, SEXPREC, Rf_xlength, R_xlen_t
 #include "cpp11/protect.hpp"  // for stop, protect, safe, protect::function

--- a/inst/include/cpp11/as.hpp
+++ b/inst/include/cpp11/as.hpp
@@ -31,7 +31,7 @@ struct is_smart_ptr<std::weak_ptr<T>> : std::true_type {};
 
 template <typename T, typename R = void>
 using enable_if_constructible_from_sexp =
-    enable_if_t<!is_smart_ptr<T>::value && // workaround for gcc 4.8
+    enable_if_t<!is_smart_ptr<T>::value &&  // workaround for gcc 4.8
                     std::is_class<T>::value && std::is_constructible<T, SEXP>::value,
                 R>;
 

--- a/inst/include/cpp11/as.hpp
+++ b/inst/include/cpp11/as.hpp
@@ -32,7 +32,7 @@ struct is_smart_ptr<std::weak_ptr<T>> : std::true_type {};
 template <typename T, typename R = void>
 using enable_if_constructible_from_sexp =
     enable_if_t<!is_smart_ptr<T>::value && // workaround for gcc 4.8
-                  std::is_class<T>::value && std::is_constructible<T, SEXP>::value,
+                    std::is_class<T>::value && std::is_constructible<T, SEXP>::value,
                 R>;
 
 template <typename T, typename R = void>

--- a/inst/include/cpp11/as.hpp
+++ b/inst/include/cpp11/as.hpp
@@ -1,10 +1,10 @@
 #pragma once
 
-#include <cmath>              // for modf
-#include <initializer_list>   // for initializer_list
-#include <string>             // for string, basic_string
-#include <type_traits>        // for decay, enable_if, is_same, is_convertible
-#include <memory>             // for std::shared_ptr, std::weak_ptr, std::unique_ptr
+#include <cmath>            // for modf
+#include <initializer_list> // for initializer_list
+#include <string>           // for string, basic_string
+#include <type_traits>      // for decay, enable_if, is_same, is_convertible
+#include <memory>           // for std::shared_ptr, std::weak_ptr, std::unique_ptr
 
 #include "cpp11/R.hpp"        // for SEXP, SEXPREC, Rf_xlength, R_xlen_t
 #include "cpp11/protect.hpp"  // for stop, protect, safe, protect::function

--- a/inst/include/cpp11/as.hpp
+++ b/inst/include/cpp11/as.hpp
@@ -1,10 +1,10 @@
 #pragma once
 
-#include <cmath>            // for modf
-#include <initializer_list> // for initializer_list
-#include <string>           // for string, basic_string
-#include <type_traits>      // for decay, enable_if, is_same, is_convertible
-#include <memory>           // for std::shared_ptr, std::weak_ptr, std::unique_ptr
+#include <cmath>             // for modf
+#include <initializer_list>  // for initializer_list
+#include <string>            // for string, basic_string
+#include <type_traits>       // for decay, enable_if, is_same, is_convertible
+#include <memory>            // for std::shared_ptr, std::weak_ptr, std::unique_ptr
 
 #include "cpp11/R.hpp"        // for SEXP, SEXPREC, Rf_xlength, R_xlen_t
 #include "cpp11/protect.hpp"  // for stop, protect, safe, protect::function

--- a/inst/include/cpp11/as.hpp
+++ b/inst/include/cpp11/as.hpp
@@ -20,13 +20,13 @@ template <typename T>
 struct is_smart_ptr : std::false_type {};
 
 template <typename T>
-struct is_smart_ptr<std::shared_ptr<T>> : std::true_type{};
+struct is_smart_ptr<std::shared_ptr<T>> : std::true_type {};
 
 template <typename T>
-struct is_smart_ptr<std::unique_ptr<T>> : std::true_type{};
+struct is_smart_ptr<std::unique_ptr<T>> : std::true_type {};
 
 template <typename T>
-struct is_smart_ptr<std::weak_ptr<T>> : std::true_type{};
+struct is_smart_ptr<std::weak_ptr<T>> : std::true_type {};
 
 template <typename T, typename R = void>
 using enable_if_constructible_from_sexp =

--- a/inst/include/cpp11/as.hpp
+++ b/inst/include/cpp11/as.hpp
@@ -2,9 +2,9 @@
 
 #include <cmath>             // for modf
 #include <initializer_list>  // for initializer_list
+#include <memory>            // for std::shared_ptr, std::weak_ptr, std::unique_ptr
 #include <string>            // for string, basic_string
 #include <type_traits>       // for decay, enable_if, is_same, is_convertible
-#include <memory>            // for std::shared_ptr, std::weak_ptr, std::unique_ptr
 
 #include "cpp11/R.hpp"        // for SEXP, SEXPREC, Rf_xlength, R_xlen_t
 #include "cpp11/protect.hpp"  // for stop, protect, safe, protect::function

--- a/inst/include/cpp11/as.hpp
+++ b/inst/include/cpp11/as.hpp
@@ -16,6 +16,18 @@ using enable_if_t = typename std::enable_if<C, R>::type;
 template <typename T>
 using decay_t = typename std::decay<T>::type;
 
+template <typename T>
+struct is_smart_ptr : std::false_type {};;
+
+template <typename T>
+struct is_smart_ptr<std::shared_ptr<T>> : std::true_type{};
+
+template <typename T>
+struct is_smart_ptr<std::unique_ptr<T>> : std::true_type{};
+
+template <typename T>
+struct is_smart_ptr<std::weak_ptr<T>> : std::true_type{};
+
 template <typename T, typename R = void>
 using enable_if_constructible_from_sexp =
     enable_if_t<std::is_class<T>::value && std::is_constructible<T, SEXP>::value, R>;
@@ -62,7 +74,7 @@ inline bool is_convertable_without_loss_to_integer(double value) {
   return std::modf(value, &int_part) == 0.0;
 }
 
-template <typename T>
+template <typename T, typename = enable_if_t<!is_smart_ptr<T>::value>>
 enable_if_constructible_from_sexp<T, T> as_cpp(SEXP from) {
   return T(from);
 }

--- a/inst/include/cpp11/as.hpp
+++ b/inst/include/cpp11/as.hpp
@@ -31,11 +31,9 @@ struct is_smart_ptr<std::weak_ptr<T>> : std::true_type {};
 
 template <typename T, typename R = void>
 using enable_if_constructible_from_sexp =
-    enable_if_t<
-      !is_smart_ptr<T>::value && // workaround for gcc 4.8
-      std::is_class<T>::value && std::is_constructible<T, SEXP>::value,
-      R
-    >;
+    enable_if_t<!is_smart_ptr<T>::value && // workaround for gcc 4.8
+                  std::is_class<T>::value && std::is_constructible<T, SEXP>::value,
+                R>;
 
 template <typename T, typename R = void>
 using enable_if_is_sexp = enable_if_t<std::is_same<T, SEXP>::value, R>;

